0.3.4 Doesn't work on Ubuntu 14.04

./DriveShare: error while loading shared libraries: libgtk-x11-2.0.so.0: cannot open shared object file: No such file or directory

Latest DEB package still doesn't run on Ubuntu 14.04.
Leaving here the errors so you can iron out some bugs.
Next I'll try to install manually.

[6666:1210/001839:ERROR:browser_main_loop.cc(189)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
A JavaScript error occurred in the main process
Uncaught Exception:
Error: Cannot find module 'fs-jetpack'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:286:25)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/opt/DriveShare/resources/app.asar/lib/electron_boilerplate/window_state.js:8:15)
    at Module._compile (module.js:430:26)
    at Object.Module._extensions..js (module.js:448:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)

--- a/app/client.js
+++ b/app/client.js
@@ -65,7 +65,7 @@ var setup = new Vue({
   el: '#setup',
   data: {
     title: 'Welcome to DriveShare',
-    working: installer._platform !== 'linux',
+    working: true,
     status: '',
     linux: installer._platform === 'linux',
     password: '',
@@ -92,10 +92,9 @@ var setup = new Vue({
         self.working = false;
 
         if (self.linux) {
-          // avoid passing error through on gnu/linux as the user's password
-          // may be shown in the failed command
-          self.error = 'Sorry, the initial setup failed. ' +
-                       'Did you enter the correct password?';
+          self.error = 'It looks like you are missing the required ' +
+                       'dependencies for GNU/Linux. Please install ' +
+                       'DriveShare using our Debian package.';
         } else {
           self.error = err.message;
         }
@@ -110,6 +109,10 @@ var setup = new Vue({
     },
     reload: function() {
       location.reload();
+    },
+    openReleases: function() {
+      shell.openExternal('https://github.com/storj/driveshare-gui/releases');
+      app.quit();
     }
   },
   created: function() {

--- a/app/client.js
+++ b/app/client.js
@@ -124,10 +124,7 @@ var setup = new Vue({
           self.status = err.message;
         }
 
-        if (!self.linux) {
-          self.setup();
-        }
-
+        self.setup();
         $('#setup').modal('show');
       }
     });

--- a/app/driveshare.html
+++ b/app/driveshare.html
@@ -138,18 +138,6 @@
           <p v-if="working">Performing first time initialization. This can take up to several minutes, hang tight...</p>
           <p v-if="working">{{status}}</p>
           <p v-if="!working && error" class="error">{{error}}</p>
-          <p v-if="linux && !working && !error">Your password is required for first time setup, please input your user password:</p>
-          <form v-if="linux && !working && !error" class="row" v-on:submit="setup">
-            <div class="col-lg-4 col-lg-offset-4">
-              <div class="input-group passwd-linux">
-                <input class="form-control input-lg" type="password" placeholder="Password" autofocus v-model="password">
-              </div>
-              <br>
-              <div class="btn-group">
-                <button type="button" class="btn btn-blue passwd-linux-btn" v-on:click="setup">Submit</button>
-              </div>
-            </div>
-          </form>
           <div v-if="working" class="spinner spinner-modal">
             <div class="rect1"></div>
             <div class="rect2"></div>
@@ -160,7 +148,8 @@
         </div>
         <div v-if="error" class="modal-footer">
           <div class="btn-group">
-            <button type="button" class="btn btn-blue" data-dismiss="modal" v-on:click="reload">Try Again</button>
+            <button v-if="linux" type="button" class="btn btn-blue" v-on:click="openReleases">Download</button>
+            <button v-if="!linux" type="button" class="btn btn-blue" data-dismiss="modal" v-on:click="reload">Try Again</button>
           </div>
         </div>
       </div>

--- a/app/lib/installer.js
+++ b/app/lib/installer.js
@@ -35,7 +35,7 @@ function DataServInstaller(datadir) {
 
   this._targets = {
     linux: {
-      install: this._installGnuLinux.bind(this),
+      install: null, // this is handled by debian package management
       check: this._checkGnuLinux.bind(this),
       path: 'dataserv-client'
     },
@@ -58,9 +58,8 @@ inherits(DataServInstaller, EventEmitter);
 /**
  * Initializes the installer and begins emitting events
  * #install
- * @param {String} password - gnu/linux only
  */
-DataServInstaller.prototype.install = function(password) {
+DataServInstaller.prototype.install = function() {
   if (Object.keys(this._targets).indexOf(this._platform) === -1) {
     return this.emit('error', new Error('This platform is not supported'));
   }
@@ -76,7 +75,14 @@ DataServInstaller.prototype.install = function(password) {
     }
 
     if (!installed) {
-      return platform.install(password);
+      if (self._platform === 'linux') {
+        return self.emit(
+          'error',
+          new Error('Dependencies must be installed via APT on GNU/Linux')
+        );
+      }
+
+      return platform.install();
     }
 
     self._logger.append('The dataserv-client is installed!');
@@ -103,74 +109,6 @@ DataServInstaller.prototype.check = function(callback) {
  */
 DataServInstaller.prototype.getDataServClientPath = function() {
   return this._targets[this._platform].path;
-};
-
-/**
- * Installs dataserv-client on gnu+linux systems
- * #_installGnuLinux
- * @param {String} passwd
- */
-DataServInstaller.prototype._installGnuLinux = function(passwd) {
-  var self = this;
-  var sudo = 'echo ' + passwd + ' | sudo -S ';
-
-  var aptdeps = [
-    'python-pip',
-    'python-dev',
-    'graphviz',
-    'libgraphviz-dev',
-    'pkg-config',
-    'gcc'
-  ];
-  var pipdeps = [
-    'dataserv-client'
-  ];
-  var pygraphviz = sudo + 'pip install pygraphviz ' +
-                   '--install-option="--include-path=/usr/include/graphviz" ' +
-                   '--install-option="--library-path=/usr/lib/graphviz/"';
-
-  var aptinstall = sudo + 'apt-get install ' + aptdeps.join(' ') + ' -y';
-  var pipinstall = sudo + 'pip install ' + pipdeps.join(' ');
-
-  this._checkPythonPipGnuLinux(function(err, installed) {
-    if (err) {
-      return self.emit('error', err);
-    }
-
-    if (!installed) {
-      self._logger.append('Installing dependencies...');
-
-      return exec(aptinstall, function(err, stdout) {
-        if (err) {
-          return self.emit('error', err);
-        }
-
-        self._logger.append(stdout);
-        _installDataservClient();
-      });
-    }
-
-    _installDataservClient();
-  });
-
-  function _installDataservClient() {
-    self._logger.append('Installing dataserv-client...');
-    exec(pygraphviz, function(err, stdout) {
-      if (err) {
-        return self.emit('error', err);
-      }
-
-      self._logger.append(stdout);
-      exec(pipinstall, function(err, stdout) {
-        if (err) {
-          return self.emit('error', err);
-        }
-
-        self._logger.append(stdout);
-        self.emit('end');
-      });
-    });
-  }
 };
 
 /**
@@ -233,21 +171,6 @@ DataServInstaller.prototype._checkMacintosh = function(callback) {
 DataServInstaller.prototype._checkWindows = function(callback) {
   fs.exists(this.getDataServClientPath(), function(exists) {
     callback(null, exists);
-  });
-};
-
-/**
- * Check if python-pip is installed on gnu+linux
- * #_checkPythonPipGnuLinux
- * @param {Function} callback
- */
-DataServInstaller.prototype._checkPythonPipGnuLinux = function(callback) {
-  exec('which pip', function(err, stdout, stderr) {
-    if (err || stderr) {
-      return callback(null, false);
-    }
-
-    callback(null, true);
   });
 };
 

--- a/app/test/unit/installer.unit.js
+++ b/app/test/unit/installer.unit.js
@@ -47,13 +47,28 @@ describe('DataServInstaller', function() {
       installer.install();
     });
 
-    it('should not install if already installed', function(done) {
+    it('should emit error on gnu/linux if not installed', function(done) {
       var installer = new Installer(tmpdir);
       installer._platform = 'linux';
       var _check = sinon.stub(installer._targets.linux, 'check', function(cb) {
+        cb(null, false);
+      });
+      installer.once('error', function(err) {
+        expect(err.message).to.equal(
+          'Dependencies must be installed via APT on GNU/Linux'
+        );
+        done();
+      });
+      installer.install();
+    });
+
+    it('should not install if already installed', function(done) {
+      var installer = new Installer(tmpdir);
+      installer._platform = 'win32';
+      var _check = sinon.stub(installer._targets.win32, 'check', function(cb) {
         cb(null, true);
       });
-      var _install = sinon.stub(installer._targets.linux, 'install');
+      var _install = sinon.stub(installer._targets.win32, 'install');
       installer.once('end', function() {
         expect(_install.called).to.equal(false);
         _check.restore();
@@ -65,11 +80,11 @@ describe('DataServInstaller', function() {
 
     it('should call the private install method', function(done) {
       var installer = new Installer(tmpdir);
-      installer._platform = 'linux';
-      var _check = sinon.stub(installer._targets.linux, 'check', function(cb) {
+      installer._platform = 'darwin';
+      var _check = sinon.stub(installer._targets.darwin, 'check', function(cb) {
         cb(null, false);
       });
-      var _inst = sinon.stub(installer._targets.linux, 'install', function() {
+      var _inst = sinon.stub(installer._targets.darwin, 'install', function() {
         installer.emit('end');
       });
       installer.once('end', function() {
@@ -131,158 +146,6 @@ describe('DataServInstaller', function() {
       installer._platform = 'linux';
       var dspath = installer.getDataServClientPath();
       expect(dspath).to.equal('dataserv-client');
-    });
-
-  });
-
-  describe('#_installGnuLinux', function() {
-
-    it('should check if pip is installed', function() {
-      var installer = new Installer(tmpdir);
-      var _check = sinon.stub(installer, '_checkPythonPipGnuLinux');
-      installer._installGnuLinux();
-      expect(_check.called).to.equal(true);
-      _check.restore();
-    });
-
-    it('should emit error if pip check fails', function(done) {
-      var installer = new Installer(tmpdir);
-      var _check = sinon.stub(
-        installer,
-        '_checkPythonPipGnuLinux',
-        function(callback) {
-          callback(new Error('Failed to check pip'));
-        }
-      );
-      installer.once('error', function(err) {
-        expect(err.message).to.equal('Failed to check pip');
-        _check.restore();
-        done();
-      });
-      installer._installGnuLinux();
-    });
-
-    it('should install pip if it is not installed', function(done) {
-      var _exec = sinon.stub().callsArg(1);
-      var Installer = proxyquire('../../lib/installer', {
-        child_process: {
-          exec: _exec
-        }
-      });
-      var installer = new Installer(tmpdir);
-      var _check = sinon.stub(
-        installer,
-        '_checkPythonPipGnuLinux',
-        function(callback) {
-          callback(null, false);
-        }
-      );
-      installer.once('end', function() {
-        expect(_exec.callCount).to.equal(3);
-        _check.restore();
-        done();
-      });
-      installer._installGnuLinux();
-    });
-
-    it('should emit error if pip install fails', function(done) {
-      var _exec = sinon.stub().callsArgWith(1, new Error('Failed'));
-      var Installer = proxyquire('../../lib/installer', {
-        child_process: {
-          exec: _exec
-        }
-      });
-      var installer = new Installer(tmpdir);
-      var _check = sinon.stub(
-        installer,
-        '_checkPythonPipGnuLinux',
-        function(callback) {
-          callback(null, false);
-        }
-      );
-      installer.once('error', function(err) {
-        expect(err.message).to.equal('Failed');
-        _check.restore();
-        done();
-      });
-      installer._installGnuLinux();
-    });
-
-    it('should install dataserv client', function(done) {
-      var _exec = sinon.stub().callsArg(1);
-      var Installer = proxyquire('../../lib/installer', {
-        child_process: {
-          exec: _exec
-        }
-      });
-      var installer = new Installer(tmpdir);
-      var _check = sinon.stub(
-        installer,
-        '_checkPythonPipGnuLinux',
-        function(callback) {
-          callback(null, true);
-        }
-      );
-      installer.once('end', function() {
-        expect(_exec.callCount).to.equal(2);
-        _check.restore();
-        done();
-      });
-      installer._installGnuLinux();
-    });
-
-    it('should emit error is pygraphviz install fails', function(done) {
-      var _exec = sinon.stub().callsArgWith(1, new Error('Failed'));
-      var Installer = proxyquire('../../lib/installer', {
-        child_process: {
-          exec: _exec
-        }
-      });
-      var installer = new Installer(tmpdir);
-      var _check = sinon.stub(
-        installer,
-        '_checkPythonPipGnuLinux',
-        function(callback) {
-          callback(null, true);
-        }
-      );
-      installer.once('error', function(err) {
-        expect(err.message).to.equal('Failed');
-        _check.restore();
-        done();
-      });
-      installer._installGnuLinux();
-    });
-
-    it('should emit error if dataserv install fails', function(done) {
-      var called = false;
-      var _exec = function(cmd, callback) {
-        if (called) {
-          callback(new Error('Failed'));
-        } else {
-          called = true;
-          callback();
-        }
-      };
-      var Installer = proxyquire('../../lib/installer', {
-        child_process: {
-          exec: _exec
-        }
-      });
-      var installer = new Installer(tmpdir);
-      var _check = sinon.stub(
-        installer,
-        '_checkPythonPipGnuLinux',
-        function(callback) {
-          callback(null, true);
-        }
-      );
-      installer.once('error', function(err) {
-        expect(err.message).to.equal('Failed');
-        _check.restore();
-        done();
-      });
-      installer._installGnuLinux();
     });
 
   });
@@ -421,51 +284,6 @@ describe('DataServInstaller', function() {
       installer._checkWindows(function(err, exists) {
         expect(err).to.equal(null);
         expect(exists).to.equal(true);
-        done();
-      });
-    });
-
-  });
-
-  describe('#_checkPythonPipGnuLinux', function() {
-
-    it('should return false if `which` fails', function(done) {
-      var Installer = proxyquire('../../lib/installer', {
-        child_process: {
-          exec: sinon.stub().callsArgWith(1, new Error('Failed'))
-        }
-      });
-      var installer = new Installer(tmpdir);
-      installer._checkPythonPipGnuLinux(function(err, installed) {
-        expect(installed).to.equal(false);
-        done();
-      });
-    });
-
-    it('should return false if `which` report missing', function(done) {
-      var Installer = proxyquire('../../lib/installer', {
-        child_process: {
-          exec: sinon.stub().callsArgWith(1, null, '', 'Not found')
-        }
-      });
-      var installer = new Installer(tmpdir);
-      installer._checkPythonPipGnuLinux(function(err, installed) {
-        expect(err).to.equal(null);
-        expect(installed).to.equal(false);
-        done();
-      });
-    });
-
-    it('should return true if `which` finds path', function(done) {
-      var Installer = proxyquire('../../lib/installer', {
-        child_process: {
-          exec: sinon.stub().callsArgWith(1, null, '/some/path', '')
-        }
-      });
-      var installer = new Installer(tmpdir);
-      installer._checkPythonPipGnuLinux(function(err, installed) {
-        expect(err).to.equal(null);
-        expect(installed).to.equal(true);
         done();
       });
     });

--- a/resources/linux/DEBIAN/control
+++ b/resources/linux/DEBIAN/control
@@ -5,3 +5,4 @@ Priority: optional
 Architecture: amd64
 Installed-Size: {{size}}
 Description: {{description}}
+Depends: python-pip, python-dev, graphviz, libgraphviz-dev, pkg-config, gcc

--- a/resources/linux/DEBIAN/postinst
+++ b/resources/linux/DEBIAN/postinst
@@ -1,0 +1,5 @@
+#!/bin/sh
+
+pip install pygraphviz --install-option="--include-path=/usr/include/graphviz" --install-option="--library-path=/usr/lib/graphviz/"
+pip install dataserv-client
+ln -s /opt/driveshare/driveshare /usr/local/bin/driveshare

--- a/tasks/release_linux.js
+++ b/tasks/release_linux.js
@@ -6,6 +6,7 @@ var childProcess = require('child_process');
 var jetpack = require('fs-jetpack');
 var asar = require('asar');
 var utils = require('./utils');
+var fs = require('fs');
 
 var projectDir;
 var releasesDir;
@@ -16,97 +17,113 @@ var readyAppDir;
 var manifest;
 
 var init = function () {
-    projectDir = jetpack;
-    tmpDir = projectDir.dir('./tmp', { empty: true });
-    releasesDir = projectDir.dir('./releases');
-    manifest = projectDir.read('app/package.json', 'json');
-    packName = manifest.name + '_' + manifest.version;
-    packDir = tmpDir.dir(packName);
-    readyAppDir = packDir.cwd('opt', manifest.name);
+  projectDir = jetpack;
+  tmpDir = projectDir.dir('./tmp', { empty: true });
+  releasesDir = projectDir.dir('./releases');
+  manifest = projectDir.read('app/package.json', 'json');
+  packName = manifest.name + '_' + manifest.version;
+  packDir = tmpDir.dir(packName);
+  readyAppDir = packDir.cwd('opt', manifest.name);
 
-    return Q();
+  return Q();
 };
 
 var copyRuntime = function () {
-    return projectDir.copyAsync('node_modules/electron-prebuilt/dist', readyAppDir.path(), { overwrite: true });
+  return projectDir.copyAsync(
+    'node_modules/electron-prebuilt/dist',
+    readyAppDir.path(),
+    { overwrite: true }
+  );
 };
 
 var packageBuiltApp = function () {
-    var deferred = Q.defer();
+  var deferred = Q.defer();
 
-    asar.createPackage(projectDir.path('build'), readyAppDir.path('resources/app.asar'), function() {
-        deferred.resolve();
-    });
+  asar.createPackage(
+    projectDir.path('build'),
+    readyAppDir.path('resources/app.asar'),
+    function() {
+      deferred.resolve();
+    }
+  );
 
-    return deferred.promise;
+  return deferred.promise;
 };
 
 var finalize = function () {
-    // Create .desktop file from the template
-    var desktop = projectDir.read('resources/linux/app.desktop');
-    desktop = utils.replace(desktop, {
-        name: manifest.name,
-        productName: manifest.productName,
-        description: manifest.description,
-        version: manifest.version,
-        author: manifest.author
-    });
-    packDir.write('usr/share/applications/' + manifest.name + '.desktop', desktop);
+  // Create .desktop file from the template
+  var desktop = projectDir.read('resources/linux/app.desktop');
+  desktop = utils.replace(desktop, {
+    name: manifest.name,
+    productName: manifest.productName,
+    description: manifest.description,
+    version: manifest.version,
+    author: manifest.author
+  });
+  packDir.write(
+    'usr/share/applications/' + manifest.name + '.desktop', desktop
+  );
 
-    // Copy icon
-    projectDir.copy('resources/icon.png', readyAppDir.path('icon.png'));
+  // Copy icon
+  projectDir.copy('resources/icon.png', readyAppDir.path('icon.png'));
 
-    return Q();
+  return Q();
 };
 
 var renameApp = function() {
-    return readyAppDir.renameAsync("electron", manifest.name);
+  return readyAppDir.renameAsync('electron', manifest.name);
 };
 
 var packToDebFile = function () {
-    var deferred = Q.defer();
+  var deferred = Q.defer();
 
-    var debFileName = packName + '_amd64.deb';
-    var debPath = releasesDir.path(debFileName);
+  var debFileName = packName + '_amd64.deb';
+  var debPath = releasesDir.path(debFileName);
 
-    gulpUtil.log('Creating DEB package...');
+  gulpUtil.log('Creating DEB package...');
 
-    // Counting size of the app in KiB
-    var appSize = Math.round(readyAppDir.inspectTree('.').size / 1024);
+  // Counting size of the app in KiB
+  var appSize = Math.round(readyAppDir.inspectTree('.').size / 1024);
 
-    // Preparing debian control file
-    var control = projectDir.read('resources/linux/DEBIAN/control');
-    control = utils.replace(control, {
-        name: manifest.name,
-        description: manifest.description,
-        version: manifest.version,
-        author: manifest.author,
-        size: appSize
+  // Preparing debian control file
+  var control = projectDir.read('resources/linux/DEBIAN/control');
+  control = utils.replace(control, {
+    name: manifest.name,
+    description: manifest.description,
+    version: manifest.version,
+    author: manifest.author,
+    size: appSize
+  });
+  packDir.write('DEBIAN/control', control);
+  projectDir.copy(
+    'resources/linux/DEBIAN/postinst',
+    packDir.path('DEBIAN/postinst')
+  );
+  fs.chmodSync(packDir.path('DEBIAN/postinst'), '755');
+
+  // Build the package...
+  childProcess.exec(
+    'fakeroot dpkg-deb -Zxz --build ' + packDir.path() + ' ' + debPath,
+    function (error, stdout, stderr) {
+      if (error || stderr) {
+        console.log('ERROR while building DEB package:');
+        console.log(error);
+        console.log(stderr);
+      } else {
+        gulpUtil.log('DEB package ready!', debPath);
+      }
+      deferred.resolve();
     });
-    packDir.write('DEBIAN/control', control);
 
-    // Build the package...
-    childProcess.exec('fakeroot dpkg-deb -Zxz --build ' + packDir.path() + ' ' + debPath,
-        function (error, stdout, stderr) {
-            if (error || stderr) {
-                console.log("ERROR while building DEB package:");
-                console.log(error);
-                console.log(stderr);
-            } else {
-                gulpUtil.log('DEB package ready!', debPath);
-            }
-            deferred.resolve();
-        });
-
-    return deferred.promise;
+  return deferred.promise;
 };
 
 var cleanClutter = function () {
-    return tmpDir.removeAsync('.');
+  return tmpDir.removeAsync('.');
 };
 
 module.exports = function () {
-    return init()
+  return init()
     .then(copyRuntime)
     .then(packageBuiltApp)
     .then(finalize)

